### ISSUE TITLE
Alpha blending can break in the OpenGL2 backend as not enough state is cleared

### ISSUE THREAD

**Issue by AsuMagic** [ocornut/imgui#3000]
**Version/Branch of Dear ImGui:**

Version: 1.74
Branch: master

**Back-end/Renderer/Compiler/OS**

Back-ends: `imgui_impl_opengl2.cpp` + `imgui_impl_sdl.cpp`
Operating System: Linux 64-bit

**My Issue/Question:**

I am working on a codebase that uses Irrlicht for rendering. I noticed that dear imgui alpha blending breaks when the imgui rendering is performed after certain Irrlicht draw calls only, such as calls to Irrlicht's `draw2DImageBatch`.

After digging up a bit, I found the culprit OpenGL calls in Irrlicht I found thanks to apitrace:
```cpp
glTexEnvf(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_COMBINE_EXT);
glTexEnvf(GL_TEXTURE_ENV, GL_COMBINE_ALPHA_EXT, GL_REPLACE);
glTexEnvf(GL_TEXTURE_ENV, GL_SOURCE0_ALPHA_EXT, GL_TEXTURE);
```

I managed to fix this issue by calling the following before the `ImGui_ImplOpenGL2_RenderDrawData` call.
```cpp
glTexEnvf(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_MODULATE);
```

**Screenshots/Video**

Alpha blending OK

![Correct alpha blending](https://user-images.githubusercontent.com/8986152/72993416-a43d8b00-3df5-11ea-8b80-79218732779b.png)

Alpha blending not OK

![Incorrect alpha blending](https://user-images.githubusercontent.com/8986152/72993638-039b9b00-3df6-11ea-828b-6e850f2c3e98.png)

**Standalone, minimal, complete and verifiable example:**

```cpp
#include "imgui.h"
#include "imgui_impl_sdl.h"
#include "imgui_impl_opengl2.h"
#include <stdio.h>
#include <SDL.h>
#include <SDL_opengl.h>

// Main code
int main(int, char**)
{
    if (SDL_Init(SDL_INIT_VIDEO | SDL_INIT_TIMER | SDL_INIT_GAMECONTROLLER) != 0)
    {
        printf("Error: %s\n", SDL_GetError());
        return -1;
    }

    SDL_GL_SetAttribute(SDL_GL_DOUBLEBUFFER, 1);
    SDL_GL_SetAttribute(SDL_GL_DEPTH_SIZE, 24);
    SDL_GL_SetAttribute(SDL_GL_STENCIL_SIZE, 8);
    SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 2);
    SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 2);
    SDL_WindowFlags window_flags = (SDL_WindowFlags)(SDL_WINDOW_OPENGL | SDL_WINDOW_RESIZABLE | SDL_WINDOW_ALLOW_HIGHDPI);
    SDL_Window* window = SDL_CreateWindow("Dear ImGui SDL2+OpenGL example", SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED, 1280, 720, window_flags);
    SDL_GLContext gl_context = SDL_GL_CreateContext(window);
    SDL_GL_MakeCurrent(window, gl_context);
    SDL_GL_SetSwapInterval(1);

    IMGUI_CHECKVERSION();
    ImGui::CreateContext();
    ImGuiIO& io = ImGui::GetIO(); (void)io;

    ImGui::StyleColorsDark();

    ImGui_ImplSDL2_InitForOpenGL(window, gl_context);
    ImGui_ImplOpenGL2_Init();

    bool show_demo_window = true;
    ImVec4 clear_color = ImVec4(0.45f, 0.55f, 0.60f, 1.00f);

    bool done = false;
    while (!done)
    {
        SDL_Event event;
        while (SDL_PollEvent(&event))
        {
            ImGui_ImplSDL2_ProcessEvent(&event);
            if (event.type == SDL_QUIT)
                done = true;
        }

        ImGui_ImplOpenGL2_NewFrame();
        ImGui_ImplSDL2_NewFrame(window);
        ImGui::NewFrame();

        if (show_demo_window)
            ImGui::ShowDemoWindow(&show_demo_window);

        ImGui::Render();
        glViewport(0, 0, (int)io.DisplaySize.x, (int)io.DisplaySize.y);
        glClearColor(clear_color.x, clear_color.y, clear_color.z, clear_color.w);
        glClear(GL_COLOR_BUFFER_BIT);

        // Mess up the alpha blending state
        glTexEnvf(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_COMBINE_EXT);
        glTexEnvf(GL_TEXTURE_ENV, GL_COMBINE_ALPHA_EXT, GL_REPLACE);
        glTexEnvf(GL_TEXTURE_ENV, GL_SOURCE0_ALPHA_EXT, GL_TEXTURE);

        // Uncomment the line below to fix the state
        // glTexEnvf(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_MODULATE);

        ImGui_ImplOpenGL2_RenderDrawData(ImGui::GetDrawData());
        SDL_GL_SwapWindow(window);
    }

    ImGui_ImplOpenGL2_Shutdown();
    ImGui_ImplSDL2_Shutdown();
    ImGui::DestroyContext();

    SDL_GL_DeleteContext(gl_context);
    SDL_DestroyWindow(window);
    SDL_Quit();
}
```

**Comment by ocornut** [comment]
Hello Asu,

Thank you for the detailed bug report!

I have now pushed a fix in f6da5000bfaf3a035e8d8a7db9b5150644f4630b including backup/restore of the current value.

Using `glTexEnvi()` instead of `glTexEnvf()` I think is more accurate here.
Could you confirm it is working out for you?

**Comment by AsuMagic** [comment]
It does, thanks!
